From 3833bc1fd7f8b47cb9e5609454e0753ad104dcd1 Mon Sep 17 00:00:00 2001
From: Thiraviyam Mariyappan <tmariyap@qti.qualcomm.com>
Date: Thu, 22 May 2025 14:22:56 +0530
Subject: [PATCH 2/2] QSDK:DP:wifi: Allocate cachable memory and Avoid dma map
 unmap for IPQ5424

Allocate cachable memory for all the rings in IO coherency supported
platforms for the rings to enable fast data access and offload
related CMO operations to hardware.

Additionally, DMA map and unmap operation is not required for
IO coherency supported platform. So, avoid dma map coherent and dma
unmap coherent in driver.

Orginal patches:
985-01-QSDK-DP-wifi-Avoid-CMO-operation-for-IPQ5424.patch
985-02-QSDK-DP-wifi-Allocate-cachable-memory-for-IPQ5424.patch
985-03-QSDK-DP-wifi-Avoid-dma-map-unmap-for-IPQ5424.patch

Signed-off-by: Thiraviyam Mariyappan <tmariyap@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/ce.c          | 65 +++++++-------
 drivers/net/wireless/ath/ath12k/core.h        | 31 +++++++
 drivers/net/wireless/ath/ath12k/dbring.c      | 22 +++--
 drivers/net/wireless/ath/ath12k/dp.c          | 84 ++++++++++---------
 drivers/net/wireless/ath/ath12k/dp_mon.c      | 24 +++---
 drivers/net/wireless/ath/ath12k/dp_rx.c       | 32 ++++---
 drivers/net/wireless/ath/ath12k/hal.c         | 19 +++--
 drivers/net/wireless/ath/ath12k/htc.c         | 11 ++-
 drivers/net/wireless/ath/ath12k/mac.c         | 19 +++--
 drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c | 53 +++++++-----
 drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c | 51 +++++++----
 .../net/wireless/ath/ath12k/wifi7/hal_rx.c    |  8 +-
 drivers/net/wireless/ath/ath12k/wmi.c         |  2 +-
 13 files changed, 262 insertions(+), 159 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/ce.c b/drivers/net/wireless/ath/ath12k/ce.c
index a9fe332..de17a72 100644
--- a/drivers/net/wireless/ath/ath12k/ce.c
+++ b/drivers/net/wireless/ath/ath12k/ce.c
@@ -77,7 +77,7 @@ static int ath12k_ce_rx_post_pipe(struct ath12k_ce_pipe *pipe)
 		}
 
 		WARN_ON_ONCE(!IS_ALIGNED((unsigned long)skb->data, 4));
-
+#ifndef CONFIG_IO_COHERENCY
 		paddr = dma_map_single(ab->dev, skb->data,
 				       skb->len + skb_tailroom(skb),
 				       DMA_FROM_DEVICE);
@@ -87,15 +87,24 @@ static int ath12k_ce_rx_post_pipe(struct ath12k_ce_pipe *pipe)
 			ret = -EIO;
 			goto exit;
 		}
+#else
+		paddr = virt_to_phys(skb->data);
+		if (unlikely(!paddr)) {
+			ath12k_warn(ab, "failed to dma map ce rx buf\n");
+			dev_kfree_skb_any(skb);
+			ret = -EIO;
+			goto exit;
+		}
+#endif
 
 		ATH12K_SKB_RXCB(skb)->paddr = paddr;
 
 		ret = ath12k_ce_rx_buf_enqueue_pipe(pipe, skb, paddr);
 		if (ret) {
 			ath12k_warn(ab, "failed to enqueue rx buf: %d\n", ret);
-			dma_unmap_single(ab->dev, paddr,
-					 skb->len + skb_tailroom(skb),
-					 DMA_FROM_DEVICE);
+			ath12k_core_dma_unmap_single(ab->dev, paddr,
+						     skb->len + skb_tailroom(skb),
+						     DMA_FROM_DEVICE);
 			dev_kfree_skb_any(skb);
 			goto exit;
 		}
@@ -167,8 +176,8 @@ static void ath12k_ce_recv_process_cb(struct ath12k_ce_pipe *pipe)
 	__skb_queue_head_init(&list);
 	while (ath12k_ce_completed_recv_next(pipe, &skb, &nbytes) == 0) {
 		max_nbytes = skb->len + skb_tailroom(skb);
-		dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
-				 max_nbytes, DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
+					     max_nbytes, DMA_FROM_DEVICE);
 
 		if (unlikely(max_nbytes < nbytes)) {
 			ath12k_warn(ab, "rxed more than expected (nbytes %d, max %d)",
@@ -245,8 +254,8 @@ static void ath12k_ce_send_done_cb(struct ath12k_ce_pipe *pipe)
 		if (!skb)
 			continue;
 
-		dma_unmap_single(ab->dev, ATH12K_SKB_CB(skb)->paddr, skb->len,
-				 DMA_TO_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_CB(skb)->paddr, skb->len,
+					     DMA_TO_DEVICE);
 		dev_kfree_skb_any(skb);
 	}
 }
@@ -355,9 +364,9 @@ ath12k_ce_alloc_ring(struct ath12k_base *ab, int nentries, int desc_sz)
 	 * coherent DMA are unsupported
 	 */
 	ce_ring->base_addr_owner_space_unaligned =
-		dma_alloc_coherent(ab->dev,
-				   nentries * desc_sz + CE_DESC_RING_ALIGN,
-				   &base_addr, GFP_KERNEL);
+		ath12k_core_dma_alloc_coherent(ab->dev,
+					       nentries * desc_sz + CE_DESC_RING_ALIGN,
+					       &base_addr, GFP_KERNEL);
 	if (!ce_ring->base_addr_owner_space_unaligned) {
 		kfree(ce_ring);
 		return ERR_PTR(-ENOMEM);
@@ -563,8 +572,8 @@ static void ath12k_ce_rx_pipe_cleanup(struct ath12k_ce_pipe *pipe)
 			continue;
 
 		ring->skb[i] = NULL;
-		dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
-				 skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
+					     skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
 		dev_kfree_skb_any(skb);
 	}
 }
@@ -727,11 +736,11 @@ void ath12k_ce_free_pipes(struct ath12k_base *ab)
 		if (pipe->src_ring) {
 			desc_sz = ath12k_hal_ce_get_desc_size(hal,
 							      HAL_CE_DESC_SRC);
-			dma_free_coherent(ab->dev,
-					  pipe->src_ring->nentries * desc_sz +
-					  CE_DESC_RING_ALIGN,
-					  pipe->src_ring->base_addr_owner_space,
-					  pipe->src_ring->base_addr_ce_space);
+			ath12k_core_dma_free_coherent(ab->dev,
+						      pipe->src_ring->nentries * desc_sz +
+						      CE_DESC_RING_ALIGN,
+						      pipe->src_ring->base_addr_owner_space,
+						      pipe->src_ring->base_addr_ce_space);
 			kfree(pipe->src_ring);
 			pipe->src_ring = NULL;
 		}
@@ -739,11 +748,11 @@ void ath12k_ce_free_pipes(struct ath12k_base *ab)
 		if (pipe->dest_ring) {
 			desc_sz = ath12k_hal_ce_get_desc_size(hal,
 							      HAL_CE_DESC_DST);
-			dma_free_coherent(ab->dev,
-					  pipe->dest_ring->nentries * desc_sz +
-					  CE_DESC_RING_ALIGN,
-					  pipe->dest_ring->base_addr_owner_space,
-					  pipe->dest_ring->base_addr_ce_space);
+			ath12k_core_dma_free_coherent(ab->dev,
+						      pipe->dest_ring->nentries * desc_sz +
+						      CE_DESC_RING_ALIGN,
+						      pipe->dest_ring->base_addr_owner_space,
+						      pipe->dest_ring->base_addr_ce_space);
 			kfree(pipe->dest_ring);
 			pipe->dest_ring = NULL;
 		}
@@ -752,11 +761,11 @@ void ath12k_ce_free_pipes(struct ath12k_base *ab)
 			desc_sz =
 			  ath12k_hal_ce_get_desc_size(hal,
 						      HAL_CE_DESC_DST_STATUS);
-			dma_free_coherent(ab->dev,
-					  pipe->status_ring->nentries * desc_sz +
-					  CE_DESC_RING_ALIGN,
-					  pipe->status_ring->base_addr_owner_space,
-					  pipe->status_ring->base_addr_ce_space);
+			ath12k_core_dma_free_coherent(ab->dev,
+						      pipe->status_ring->nentries * desc_sz +
+						      CE_DESC_RING_ALIGN,
+						      pipe->status_ring->base_addr_owner_space,
+						      pipe->status_ring->base_addr_ce_space);
 			kfree(pipe->status_ring);
 			pipe->status_ring = NULL;
 		}
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 4d37f36..4dc0a6f 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1581,4 +1581,35 @@ static inline bool ath12k_hw_group_recovery_in_progress(const struct ath12k_hw_g
 	return test_bit(ATH12K_FLAG_RECOVERY, &ag->dev_flags);
 }
 
+static inline void *ath12k_core_dma_alloc_coherent(struct device *dev, size_t size,
+						   dma_addr_t *paddr, gfp_t flag)
+{
+        void *vaddr = NULL;
+#ifdef CONFIG_IO_COHERENCY
+        vaddr = kzalloc(size, flag);
+        *paddr = (dma_addr_t)virt_to_phys(vaddr);
+#else
+        vaddr = dma_alloc_coherent(dev, size, paddr, flag);
+#endif
+        return vaddr;
+}
+
+static inline void ath12k_core_dma_free_coherent(struct device *dev, size_t size,
+						 void *vaddr, dma_addr_t paddr)
+{
+#ifdef CONFIG_IO_COHERENCY
+	kfree(vaddr);
+#else
+	dma_free_coherent(dev, size, vaddr, paddr);
+#endif
+}
+
+static inline void ath12k_core_dma_unmap_single(struct device *dev, dma_addr_t dma_handle,
+						size_t size, enum dma_data_direction direction)
+{
+#ifndef CONFIG_IO_COHERENCY
+	dma_unmap_single(dev, dma_handle, size, direction);
+#endif
+}
+
 #endif /* _CORE_H_ */
diff --git a/drivers/net/wireless/ath/ath12k/dbring.c b/drivers/net/wireless/ath/ath12k/dbring.c
index bb4d71c..b558db6 100644
--- a/drivers/net/wireless/ath/ath12k/dbring.c
+++ b/drivers/net/wireless/ath/ath12k/dbring.c
@@ -58,13 +58,18 @@ static int ath12k_dbring_bufs_replenish(struct ath12k *ar,
 	ptr_unaligned = buff->payload;
 	ptr_aligned = PTR_ALIGN(ptr_unaligned, ring->buf_align);
 	ath12k_dbring_fill_magic_value(ar, ptr_aligned, ring->buf_sz);
+#ifndef CONFIG_IO_COHERENCY
 	paddr = dma_map_single(ab->dev, ptr_aligned, ring->buf_sz,
 			       DMA_FROM_DEVICE);
 
 	ret = dma_mapping_error(ab->dev, paddr);
 	if (ret)
 		goto err;
-
+#else
+	paddr = virt_to_phys(ptr_aligned);
+	if (!paddr)
+		goto err;
+#endif
 	spin_lock_bh(&ring->idr_lock);
 	buf_id = idr_alloc(&ring->bufs_idr, buff, 0, ring->bufs_max, gfp);
 	spin_unlock_bh(&ring->idr_lock);
@@ -80,8 +85,9 @@ static int ath12k_dbring_bufs_replenish(struct ath12k *ar,
 	}
 
 	buff->paddr = paddr;
-
+#ifndef CONFIG_IO_COHERENCY
 	dma_sync_single_for_device(ab->dev, paddr, ring->buf_sz, DMA_FROM_DEVICE);
+#endif
 	cookie = u32_encode_bits(ar->pdev_idx, DP_RXDMA_BUF_COOKIE_PDEV_ID) |
 		 u32_encode_bits(buf_id, DP_RXDMA_BUF_COOKIE_BUF_ID);
 
@@ -96,8 +102,8 @@ err_idr_remove:
 	idr_remove(&ring->bufs_idr, buf_id);
 	spin_unlock_bh(&ring->idr_lock);
 err_dma_unmap:
-	dma_unmap_single(ab->dev, paddr, ring->buf_sz,
-			 DMA_FROM_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, paddr, ring->buf_sz,
+				     DMA_FROM_DEVICE);
 err:
 	ath12k_hal_srng_access_end(ab, srng);
 	return ret;
@@ -343,8 +349,8 @@ int ath12k_dbring_buffer_release_event(struct ath12k_base *ab,
 		idr_remove(&ring->bufs_idr, buf_id);
 		spin_unlock_bh(&ring->idr_lock);
 
-		dma_unmap_single(ab->dev, buff->paddr, ring->buf_sz,
-				 DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, buff->paddr, ring->buf_sz,
+					     DMA_FROM_DEVICE);
 
 		if (ring->handler) {
 			vaddr_unalign = buff->payload;
@@ -380,8 +386,8 @@ void ath12k_dbring_buf_cleanup(struct ath12k *ar, struct ath12k_dbring *ring)
 	spin_lock_bh(&ring->idr_lock);
 	idr_for_each_entry(&ring->bufs_idr, buff, buf_id) {
 		idr_remove(&ring->bufs_idr, buf_id);
-		dma_unmap_single(ar->ab->dev, buff->paddr,
-				 ring->buf_sz, DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(ar->ab->dev, buff->paddr,
+					     ring->buf_sz, DMA_FROM_DEVICE);
 		kfree(buff);
 	}
 
diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 384d12f..870cd37 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -123,8 +123,8 @@ void ath12k_dp_srng_cleanup(struct ath12k_base *ab, struct dp_srng *ring)
 	if (ring->cached)
 		kfree(ring->vaddr_unaligned);
 	else
-		dma_free_coherent(ab->dev, ring->size, ring->vaddr_unaligned,
-				  ring->paddr_unaligned);
+		ath12k_core_dma_free_coherent(ab->dev, ring->size, ring->vaddr_unaligned,
+					      ring->paddr_unaligned);
 
 	ring->vaddr_unaligned = NULL;
 }
@@ -264,6 +264,7 @@ int ath12k_dp_srng_setup(struct ath12k_base *ab, struct dp_srng *ring,
 		num_entries = max_entries;
 
 	ring->size = (num_entries * entry_sz) + HAL_RING_BASE_ALIGN - 1;
+#ifndef CONFIG_IO_COHERENCY
 	if (ab->hw_params->alloc_cacheable_memory) {
 		/* Allocate the reo dst and tx completion rings from cacheable memory */
 		switch (type) {
@@ -275,6 +276,9 @@ int ath12k_dp_srng_setup(struct ath12k_base *ab, struct dp_srng *ring,
 			cached = false;
 		}
 	}
+#else
+	cached = true;
+#endif
 
 	if (cached) {
 		ring->vaddr_unaligned = kzalloc(ring->size, GFP_KERNEL);
@@ -602,8 +606,8 @@ static void ath12k_dp_scatter_idle_link_desc_cleanup(struct ath12k_base *ab)
 		if (!slist[i].vaddr)
 			continue;
 
-		dma_free_coherent(ab->dev, HAL_WBM_IDLE_SCATTER_BUF_SIZE_MAX,
-				  slist[i].vaddr, slist[i].paddr);
+		ath12k_core_dma_free_coherent(ab->dev, HAL_WBM_IDLE_SCATTER_BUF_SIZE_MAX,
+					      slist[i].vaddr, slist[i].paddr);
 		slist[i].vaddr = NULL;
 	}
 }
@@ -637,9 +641,9 @@ static int ath12k_dp_scatter_idle_link_desc_setup(struct ath12k_base *ab,
 		return -EINVAL;
 
 	for (i = 0; i < num_scatter_buf; i++) {
-		slist[i].vaddr = dma_alloc_coherent(ab->dev,
-						    HAL_WBM_IDLE_SCATTER_BUF_SIZE_MAX,
-						    &slist[i].paddr, GFP_KERNEL);
+		slist[i].vaddr = ath12k_core_dma_alloc_coherent(ab->dev,
+								HAL_WBM_IDLE_SCATTER_BUF_SIZE_MAX,
+								&slist[i].paddr, GFP_KERNEL);
 		if (!slist[i].vaddr) {
 			ret = -ENOMEM;
 			goto err;
@@ -695,10 +699,10 @@ ath12k_dp_link_desc_bank_free(struct ath12k_base *ab,
 
 	for (i = 0; i < DP_LINK_DESC_BANKS_MAX; i++) {
 		if (link_desc_banks[i].vaddr_unaligned) {
-			dma_free_coherent(ab->dev,
-					  link_desc_banks[i].size,
-					  link_desc_banks[i].vaddr_unaligned,
-					  link_desc_banks[i].paddr_unaligned);
+			ath12k_core_dma_free_coherent(ab->dev,
+						      link_desc_banks[i].size,
+						      link_desc_banks[i].vaddr_unaligned,
+						      link_desc_banks[i].paddr_unaligned);
 			link_desc_banks[i].vaddr_unaligned = NULL;
 		}
 	}
@@ -719,9 +723,9 @@ static int ath12k_dp_link_desc_bank_alloc(struct ath12k_base *ab,
 			desc_sz = last_bank_sz;
 
 		desc_bank[i].vaddr_unaligned =
-					dma_alloc_coherent(ab->dev, desc_sz,
-							   &desc_bank[i].paddr_unaligned,
-							   GFP_KERNEL);
+				ath12k_core_dma_alloc_coherent(ab->dev, desc_sz,
+							       &desc_bank[i].paddr_unaligned,
+							       GFP_KERNEL);
 		if (!desc_bank[i].vaddr_unaligned) {
 			ret = -ENOMEM;
 			goto err;
@@ -1069,8 +1073,8 @@ static void ath12k_dp_cc_cleanup(struct ath12k_base *ab)
 			if (!skb)
 				continue;
 
-			dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
-					 skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
+			ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
+						     skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
 			dev_kfree_skb_any(skb);
 		}
 	}
@@ -1098,10 +1102,10 @@ static void ath12k_dp_cc_cleanup(struct ath12k_base *ab)
 				continue;
 
 			if (tx_desc_info->skb_ext_desc) {
-				dma_unmap_single(ab->dev,
-						 ATH12K_SKB_CB(skb)->paddr_ext_desc,
-						 tx_desc_info->skb_ext_desc->len,
-						 DMA_TO_DEVICE);
+				ath12k_core_dma_unmap_single(ab->dev,
+							     ATH12K_SKB_CB(skb)->paddr_ext_desc,
+							     tx_desc_info->skb_ext_desc->len,
+							     DMA_TO_DEVICE);
 				dev_kfree_skb_any(tx_desc_info->skb_ext_desc);
 			}
 
@@ -1116,8 +1120,8 @@ static void ath12k_dp_cc_cleanup(struct ath12k_base *ab)
 					wake_up(&ar->dp.tx_empty_waitq);
 			}
 
-			dma_unmap_single(ab->dev, ATH12K_SKB_CB(skb)->paddr,
-					 skb->len, DMA_TO_DEVICE);
+			ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_CB(skb)->paddr,
+						     skb->len, DMA_TO_DEVICE);
 			dev_kfree_skb_any(skb);
 		}
 
@@ -1144,8 +1148,8 @@ static void ath12k_dp_cc_cleanup(struct ath12k_base *ab)
 		if (!dp->spt_info[i].vaddr)
 			continue;
 
-		dma_free_coherent(ab->dev, ATH12K_PAGE_SIZE,
-				  dp->spt_info[i].vaddr, dp->spt_info[i].paddr);
+		ath12k_core_dma_free_coherent(ab->dev, ATH12K_PAGE_SIZE,
+					      dp->spt_info[i].vaddr, dp->spt_info[i].paddr);
 		dp->spt_info[i].vaddr = NULL;
 	}
 
@@ -1162,17 +1166,17 @@ static void ath12k_dp_reoq_lut_cleanup(struct ath12k_base *ab)
 
 	if (dp->reoq_lut.vaddr_unaligned) {
 		ath12k_hal_write_reoq_lut_addr(ab, 0);
-		dma_free_coherent(ab->dev, dp->reoq_lut.size,
-				  dp->reoq_lut.vaddr_unaligned,
-				  dp->reoq_lut.paddr_unaligned);
+		ath12k_core_dma_free_coherent(ab->dev, dp->reoq_lut.size,
+					      dp->reoq_lut.vaddr_unaligned,
+					      dp->reoq_lut.paddr_unaligned);
 		dp->reoq_lut.vaddr_unaligned = NULL;
 	}
 
 	if (dp->ml_reoq_lut.vaddr_unaligned) {
 		ath12k_hal_write_ml_reoq_lut_addr(ab, 0);
-		dma_free_coherent(ab->dev, dp->ml_reoq_lut.size,
-				  dp->ml_reoq_lut.vaddr_unaligned,
-				  dp->ml_reoq_lut.paddr_unaligned);
+		ath12k_core_dma_free_coherent(ab->dev, dp->ml_reoq_lut.size,
+					      dp->ml_reoq_lut.vaddr_unaligned,
+					      dp->ml_reoq_lut.paddr_unaligned);
 		dp->ml_reoq_lut.vaddr_unaligned = NULL;
 	}
 }
@@ -1414,10 +1418,10 @@ static int ath12k_dp_cc_init(struct ath12k_base *ab)
 	dp->rx_ppt_base = ab->device_id * ATH12K_NUM_RX_SPT_PAGES;
 
 	for (i = 0; i < dp->num_spt_pages; i++) {
-		dp->spt_info[i].vaddr = dma_alloc_coherent(ab->dev,
-							   ATH12K_PAGE_SIZE,
-							   &dp->spt_info[i].paddr,
-							   GFP_KERNEL);
+		dp->spt_info[i].vaddr = ath12k_core_dma_alloc_coherent(ab->dev,
+								       ATH12K_PAGE_SIZE,
+								       &dp->spt_info[i].paddr,
+								       GFP_KERNEL);
 
 		if (!dp->spt_info[i].vaddr) {
 			ret = -ENOMEM;
@@ -1459,9 +1463,9 @@ static int ath12k_dp_alloc_reoq_lut(struct ath12k_base *ab,
 				    struct ath12k_reo_q_addr_lut *lut)
 {
 	lut->size =  DP_REOQ_LUT_SIZE + HAL_REO_QLUT_ADDR_ALIGN - 1;
-	lut->vaddr_unaligned = dma_alloc_coherent(ab->dev, lut->size,
-						  &lut->paddr_unaligned,
-						  GFP_KERNEL | __GFP_ZERO);
+	lut->vaddr_unaligned = ath12k_core_dma_alloc_coherent(ab->dev, lut->size,
+							      &lut->paddr_unaligned,
+							      GFP_KERNEL | __GFP_ZERO);
 	if (!lut->vaddr_unaligned)
 		return -ENOMEM;
 
@@ -1488,9 +1492,9 @@ static int ath12k_dp_reoq_lut_setup(struct ath12k_base *ab)
 	ret = ath12k_dp_alloc_reoq_lut(ab, &dp->ml_reoq_lut);
 	if (ret) {
 		ath12k_warn(ab, "failed to allocate memory for ML reoq table");
-		dma_free_coherent(ab->dev, dp->reoq_lut.size,
-				  dp->reoq_lut.vaddr_unaligned,
-				  dp->reoq_lut.paddr_unaligned);
+		ath12k_core_dma_free_coherent(ab->dev, dp->reoq_lut.size,
+					      dp->reoq_lut.vaddr_unaligned,
+					      dp->reoq_lut.paddr_unaligned);
 		dp->reoq_lut.vaddr_unaligned = NULL;
 		return ret;
 	}
diff --git a/drivers/net/wireless/ath/ath12k/dp_mon.c b/drivers/net/wireless/ath/ath12k/dp_mon.c
index 8e63a4e..017f316 100644
--- a/drivers/net/wireless/ath/ath12k/dp_mon.c
+++ b/drivers/net/wireless/ath/ath12k/dp_mon.c
@@ -2306,9 +2306,9 @@ ath12k_dp_mon_parse_status_buf(struct ath12k *ar,
 		return 0;
 	}
 
-	dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(msdu)->paddr,
-			 msdu->len + skb_tailroom(msdu),
-			 DMA_FROM_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(msdu)->paddr,
+				     msdu->len + skb_tailroom(msdu),
+				     DMA_FROM_DEVICE);
 
 	offset = packet_info->dma_length + ATH12K_MON_RX_DOT11_OFFSET;
 	if (ath12k_dp_pkt_set_pktlen(msdu, offset)) {
@@ -2469,14 +2469,18 @@ int ath12k_dp_mon_buf_replenish(struct ath12k_base *ab,
 				 PTR_ALIGN(skb->data, DP_RX_BUFFER_ALIGN_SIZE) -
 				 skb->data);
 		}
-
+#ifndef CONFIG_IO_COHERENCY
 		paddr = dma_map_single(ab->dev, skb->data,
 				       skb->len + skb_tailroom(skb),
 				       DMA_FROM_DEVICE);
 
 		if (unlikely(dma_mapping_error(ab->dev, paddr)))
 			goto fail_free_skb;
-
+#else
+		paddr = virt_to_phys(skb->data);
+		if(unlikely(!paddr))
+			goto fail_free_skb;
+#endif
 		spin_lock_bh(&buf_ring->idr_lock);
 		buf_id = idr_alloc(&buf_ring->bufs_idr, skb, 0,
 				   buf_ring->bufs_max * 3, GFP_ATOMIC);
@@ -2509,8 +2513,8 @@ fail_idr_remove:
 	idr_remove(&buf_ring->bufs_idr, buf_id);
 	spin_unlock_bh(&buf_ring->idr_lock);
 fail_dma_unmap:
-	dma_unmap_single(ab->dev, paddr, skb->len + skb_tailroom(skb),
-			 DMA_FROM_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, paddr, skb->len + skb_tailroom(skb),
+				     DMA_FROM_DEVICE);
 fail_free_skb:
 	dev_kfree_skb_any(skb);
 fail_alloc_skb:
@@ -3572,9 +3576,9 @@ int ath12k_dp_mon_srng_process(struct ath12k_pdev_dp *pdev_dp, int *budget,
 		}
 
 		rxcb = ATH12K_SKB_RXCB(skb);
-		dma_unmap_single(ab->dev, rxcb->paddr,
-				 skb->len + skb_tailroom(skb),
-				 DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, rxcb->paddr,
+					     skb->len + skb_tailroom(skb),
+					     DMA_FROM_DEVICE);
 
 		end_reason = u32_get_bits(info0, HAL_MON_DEST_INFO0_END_REASON);
 
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index f023307..cd0bfb8 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -357,12 +357,18 @@ int ath12k_dp_rx_bufs_replenish(struct ath12k_dp *dp,
 				 PTR_ALIGN(skb->data, DP_RX_BUFFER_ALIGN_SIZE) -
 				 skb->data);
 		}
-
+#ifndef CONFIG_IO_COHERENCY
 		paddr = dma_map_single(dp->dev, skb->data,
 				       skb->len + skb_tailroom(skb),
 				       DMA_FROM_DEVICE);
 		if (dma_mapping_error(dp->dev, paddr))
 			goto fail_free_skb;
+#else
+		paddr = virt_to_phys(skb->data);
+		if(unlikely(!paddr)) {
+			goto fail_free_skb;
+		}
+#endif
 
 		rx_desc = list_first_entry_or_null(used_list,
 						   struct ath12k_rx_desc_info,
@@ -389,8 +395,8 @@ int ath12k_dp_rx_bufs_replenish(struct ath12k_dp *dp,
 	goto out;
 
 fail_dma_unmap:
-	dma_unmap_single(dp->dev, paddr, skb->len + skb_tailroom(skb),
-			 DMA_FROM_DEVICE);
+	ath12k_core_dma_unmap_single(dp->dev, paddr, skb->len + skb_tailroom(skb),
+				     DMA_FROM_DEVICE);
 fail_free_skb:
 	dev_kfree_skb_any(skb);
 out:
@@ -417,8 +423,8 @@ static int ath12k_dp_rxdma_mon_buf_ring_free(struct ath12k_base *ab,
 		/* TODO: Understand where internal driver does this dma_unmap
 		 * of rxdma_buffer.
 		 */
-		dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
-				 skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, ATH12K_SKB_RXCB(skb)->paddr,
+					     skb->len + skb_tailroom(skb), DMA_FROM_DEVICE);
 		dev_kfree_skb_any(skb);
 	}
 
@@ -567,8 +573,8 @@ void ath12k_dp_rx_reo_cmd_list_cleanup(struct ath12k_base *ab)
 	spin_lock_bh(&dp->reo_cmd_lock);
 	list_for_each_entry_safe(cmd, tmp, &dp->reo_cmd_list, list) {
 		list_del(&cmd->list);
-		dma_unmap_single(ab->dev, cmd->data.paddr,
-				 cmd->data.size, DMA_BIDIRECTIONAL);
+		ath12k_core_dma_unmap_single(ab->dev, cmd->data.paddr,
+					     cmd->data.size, DMA_BIDIRECTIONAL);
 		kfree(cmd->data.vaddr);
 		kfree(cmd);
 	}
@@ -577,8 +583,8 @@ void ath12k_dp_rx_reo_cmd_list_cleanup(struct ath12k_base *ab)
 				 &dp->reo_cmd_cache_flush_list, list) {
 		list_del(&cmd_cache->list);
 		dp->reo_cmd_cache_flush_count--;
-		dma_unmap_single(ab->dev, cmd_cache->data.paddr,
-				 cmd_cache->data.size, DMA_BIDIRECTIONAL);
+		ath12k_core_dma_unmap_single(ab->dev, cmd_cache->data.paddr,
+					     cmd_cache->data.size, DMA_BIDIRECTIONAL);
 		kfree(cmd_cache->data.vaddr);
 		kfree(cmd_cache);
 	}
@@ -594,8 +600,8 @@ void ath12k_dp_reo_cmd_free(struct ath12k_dp *dp, void *ctx,
 		ath12k_warn(dp->ab, "failed to flush rx tid hw desc, tid %d status %d\n",
 			    rx_tid->tid, status);
 
-	dma_unmap_single(dp->ab->dev, rx_tid->paddr, rx_tid->size,
-			 DMA_BIDIRECTIONAL);
+	ath12k_core_dma_unmap_single(dp->ab->dev, rx_tid->paddr, rx_tid->size,
+				     DMA_BIDIRECTIONAL);
 	kfree(rx_tid->vaddr);
 	rx_tid->vaddr = NULL;
 }
@@ -656,8 +662,8 @@ void ath12k_dp_rx_tid_del_func(struct ath12k_dp *dp, void *ctx,
 
 	return;
 free_desc:
-	dma_unmap_single(ab->dev, rx_tid->paddr, rx_tid->size,
-			 DMA_BIDIRECTIONAL);
+	ath12k_core_dma_unmap_single(ab->dev, rx_tid->paddr, rx_tid->size,
+				     DMA_BIDIRECTIONAL);
 	kfree(rx_tid->vaddr);
 	rx_tid->vaddr = NULL;
 }
diff --git a/drivers/net/wireless/ath/ath12k/hal.c b/drivers/net/wireless/ath/ath12k/hal.c
index 41b75b8..531acd0 100644
--- a/drivers/net/wireless/ath/ath12k/hal.c
+++ b/drivers/net/wireless/ath/ath12k/hal.c
@@ -141,8 +141,8 @@ static int ath12k_hal_alloc_cont_rdp(struct ath12k_hal *hal)
 	size_t size;
 
 	size = sizeof(u32) * HAL_SRNG_RING_ID_MAX;
-	hal->rdp.vaddr = dma_alloc_coherent(hal->dev, size, &hal->rdp.paddr,
-					    GFP_KERNEL);
+	hal->rdp.vaddr = ath12k_core_dma_alloc_coherent(hal->dev, size, &hal->rdp.paddr,
+							GFP_KERNEL);
 	if (!hal->rdp.vaddr)
 		return -ENOMEM;
 
@@ -157,8 +157,8 @@ static void ath12k_hal_free_cont_rdp(struct ath12k_hal *hal)
 		return;
 
 	size = sizeof(u32) * HAL_SRNG_RING_ID_MAX;
-	dma_free_coherent(hal->dev, size,
-			  hal->rdp.vaddr, hal->rdp.paddr);
+	ath12k_core_dma_free_coherent(hal->dev, size,
+				      hal->rdp.vaddr, hal->rdp.paddr);
 	hal->rdp.vaddr = NULL;
 }
 
@@ -167,8 +167,8 @@ static int ath12k_hal_alloc_cont_wrp(struct ath12k_hal *hal)
 	size_t size;
 
 	size = sizeof(u32) * (HAL_SRNG_NUM_PMAC_RINGS + HAL_SRNG_NUM_DMAC_RINGS);
-	hal->wrp.vaddr = dma_alloc_coherent(hal->dev, size, &hal->wrp.paddr,
-					    GFP_KERNEL);
+	hal->wrp.vaddr = ath12k_core_dma_alloc_coherent(hal->dev, size, &hal->wrp.paddr,
+							GFP_KERNEL);
 	if (!hal->wrp.vaddr)
 		return -ENOMEM;
 
@@ -183,8 +183,8 @@ static void ath12k_hal_free_cont_wrp(struct ath12k_hal *hal)
 		return;
 
 	size = sizeof(u32) * (HAL_SRNG_NUM_PMAC_RINGS + HAL_SRNG_NUM_DMAC_RINGS);
-	dma_free_coherent(hal->dev, size,
-			  hal->wrp.vaddr, hal->wrp.paddr);
+	ath12k_core_dma_free_coherent(hal->dev, size,
+				      hal->wrp.vaddr, hal->wrp.paddr);
 	hal->wrp.vaddr = NULL;
 }
 
@@ -322,11 +322,12 @@ void *ath12k_hal_srng_dst_get_next_entry(struct ath12k_base *ab,
         if (srng->u.dst_ring.tp == srng->ring_size)
                 srng->u.dst_ring.tp = 0;
 
+#ifndef CONFIG_IO_COHERENCY
 	if (srng->flags & HAL_SRNG_FLAGS_CACHED)
 		dma_sync_single_for_cpu(ab->dev, virt_to_phys(desc),
 					(srng->entry_size * sizeof(u32)),
 					DMA_FROM_DEVICE);
-
+#endif
 	return desc;
 }
 EXPORT_SYMBOL(ath12k_hal_srng_dst_get_next_entry);
diff --git a/drivers/net/wireless/ath/ath12k/htc.c b/drivers/net/wireless/ath/ath12k/htc.c
index 607b509..177329d 100644
--- a/drivers/net/wireless/ath/ath12k/htc.c
+++ b/drivers/net/wireless/ath/ath12k/htc.c
@@ -108,13 +108,20 @@ int ath12k_htc_send(struct ath12k_htc *htc,
 	}
 
 	ath12k_htc_prepare_tx_skb(ep, skb);
-
+#ifndef CONFIG_IO_COHERENCY
 	skb_cb->paddr = dma_map_single(dev, skb->data, skb->len, DMA_TO_DEVICE);
 	ret = dma_mapping_error(dev, skb_cb->paddr);
 	if (ret) {
 		ret = -EIO;
 		goto err_credits;
 	}
+#else
+	skb_cb->paddr = virt_to_phys(skb->data);
+	if (!skb_cb->paddr) {
+		ret = -EIO;
+		goto err_credits;
+	}
+#endif
 
 	ret = ath12k_ce_send(htc->ab, skb, ep->ul_pipe_id, ep->eid);
 	if (ret)
@@ -123,7 +130,7 @@ int ath12k_htc_send(struct ath12k_htc *htc,
 	return 0;
 
 err_unmap:
-	dma_unmap_single(dev, skb_cb->paddr, skb->len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(dev, skb_cb->paddr, skb->len, DMA_TO_DEVICE);
 err_credits:
 	if (ep->tx_credit_flow_enabled) {
 		spin_lock_bh(&htc->tx_lock);
diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index e5c0634..893f502 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -9133,8 +9133,8 @@ static void ath12k_mac_tx_mgmt_free(struct ath12k *ar, int buf_id)
 	if (!msdu)
 		return;
 
-	dma_unmap_single(ar->ab->dev, ATH12K_SKB_CB(msdu)->paddr, msdu->len,
-			 DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(ar->ab->dev, ATH12K_SKB_CB(msdu)->paddr, msdu->len,
+				     DMA_TO_DEVICE);
 
 	info = IEEE80211_SKB_CB(msdu);
 	memset(&info->status, 0, sizeof(info->status));
@@ -9201,14 +9201,21 @@ static int ath12k_mac_mgmt_tx_wmi(struct ath12k *ar, struct ath12k_link_vif *arv
 			skb_put(skb, mic_len);
 		}
 	}
-
+#ifndef CONFIG_IO_COHERENCY
 	paddr = dma_map_single(ab->dev, skb->data, skb->len, DMA_TO_DEVICE);
 	if (dma_mapping_error(ab->dev, paddr)) {
 		ath12k_warn(ab, "failed to DMA map mgmt Tx buffer\n");
 		ret = -EIO;
 		goto err_free_idr;
 	}
-
+#else
+	paddr = virt_to_phys(skb->data);
+	if (!paddr) {
+		ath12k_warn(ab, "failed to DMA map mgmt Tx buffer\n");
+		ret = -EIO;
+		goto err_free_idr;
+	}
+#endif
 	skb_cb->paddr = paddr;
 
 	link_agnostic = ATH12K_SKB_CB(skb)->flags & ATH12K_SKB_MGMT_LINK_AGNOSTIC;
@@ -9223,8 +9230,8 @@ static int ath12k_mac_mgmt_tx_wmi(struct ath12k *ar, struct ath12k_link_vif *arv
 	return 0;
 
 err_unmap_buf:
-	dma_unmap_single(ab->dev, skb_cb->paddr,
-			 skb->len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, skb_cb->paddr,
+				     skb->len, DMA_TO_DEVICE);
 err_free_idr:
 	spin_lock_bh(&ar->txmgmt_idr_lock);
 	idr_remove(&ar->txmgmt_idr, buf_id);
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c b/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
index de04947..6ac50b0 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_rx.c
@@ -171,8 +171,8 @@ void ath12k_wifi7_dp_reo_cache_flush(struct ath12k_base *ab,
 	if (ret) {
 		ath12k_err(ab, "failed to send HAL_REO_CMD_FLUSH_CACHE cmd, tid %d (%d)\n",
 			   rx_tid->tid, ret);
-		dma_unmap_single(ab->dev, rx_tid->paddr, rx_tid->size,
-				 DMA_BIDIRECTIONAL);
+		ath12k_core_dma_unmap_single(ab->dev, rx_tid->paddr, rx_tid->size,
+					     DMA_BIDIRECTIONAL);
 		kfree(rx_tid->vaddr);
 		rx_tid->vaddr = NULL;
 	}
@@ -256,8 +256,8 @@ void ath12k_wifi7_dp_rx_peer_tid_delete(struct ath12k *ar,
 	if (ret) {
 		ath12k_err(ar->ab, "failed to send HAL_REO_CMD_UPDATE_RX_QUEUE cmd, tid %d (%d)\n",
 			   tid, ret);
-		dma_unmap_single(ar->ab->dev, rx_tid->paddr, rx_tid->size,
-				 DMA_BIDIRECTIONAL);
+		ath12k_core_dma_unmap_single(ar->ab->dev, rx_tid->paddr, rx_tid->size,
+					     DMA_BIDIRECTIONAL);
 		kfree(rx_tid->vaddr);
 		rx_tid->vaddr = NULL;
 	}
@@ -1093,8 +1093,9 @@ int ath12k_wifi7_dp_rx_process(struct ath12k_dp *dp, int ring_id,
 	struct sk_buff *msdu;
 	bool done = false;
 	u64 desc_va;
+#ifndef CONFIG_IO_COHERENCY
 	int valid_entries;
-
+#endif
 	__skb_queue_head_init(&msdu_list);
 
 	for (device_id = 0; device_id < ATH12K_MAX_SOCS; device_id++)
@@ -1107,6 +1108,7 @@ int ath12k_wifi7_dp_rx_process(struct ath12k_dp *dp, int ring_id,
 try_again:
 	ath12k_hal_srng_access_begin(ab, srng);
 
+#ifndef CONFIG_IO_COHERENCY
 	valid_entries = ath12k_hal_srng_dst_num_free(ab, srng, false);
 	if (unlikely(!valid_entries)) {
 		ath12k_hal_srng_access_end(ab, srng);
@@ -1114,7 +1116,7 @@ try_again:
 		return -EINVAL;
 	}
 	ath12k_hal_srng_dst_invalidate_entry(ab, srng, valid_entries);
-
+#endif
 	while ((desc = ath12k_hal_srng_dst_get_next_cached_entry(ab, srng))) {
 		struct rx_mpdu_desc *mpdu_info;
 		struct rx_msdu_desc *msdu_info;
@@ -1161,9 +1163,9 @@ try_again:
 		list_add_tail(&desc_info->list, &rx_desc_used_list[device_id]);
 
 		rxcb = ATH12K_SKB_RXCB(msdu);
-		dma_unmap_single(partner_dp->dev, rxcb->paddr,
-				 msdu->len + skb_tailroom(msdu),
-				 DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(partner_dp->dev, rxcb->paddr,
+					     msdu->len + skb_tailroom(msdu),
+					     DMA_FROM_DEVICE);
 
 		num_buffs_reaped[device_id]++;
 
@@ -1432,12 +1434,17 @@ ath12k_wifi7_dp_rx_h_defrag_reo_reinject(struct ath12k_dp *dp,
 	/* change msdu len in hal rx desc */
 	ath12k_wifi7_dp_rxdesc_set_msdu_len(ab, rx_desc, len_diff);
 
+#ifndef CONFIG_IO_COHERENCY
 	buf_paddr = dma_map_single(ab->dev, defrag_skb->data,
 				   defrag_skb->len + skb_tailroom(defrag_skb),
 				   DMA_TO_DEVICE);
 	if (dma_mapping_error(ab->dev, buf_paddr))
 		return -ENOMEM;
-
+#else
+	buf_paddr = virt_to_phys(defrag_skb->data);
+	if (!buf_paddr)
+		return -ENOMEM;
+#endif
 	spin_lock_bh(&dp->rx_desc_lock);
 	desc_info = list_first_entry_or_null(&dp->rx_desc_free_list,
 					     struct ath12k_rx_desc_info,
@@ -1524,8 +1531,8 @@ err_free_desc:
 	list_add_tail(&desc_info->list, &dp->rx_desc_free_list);
 	spin_unlock_bh(&dp->rx_desc_lock);
 err_unmap_dma:
-	dma_unmap_single(ab->dev, buf_paddr, defrag_skb->len + skb_tailroom(defrag_skb),
-			 DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, buf_paddr, defrag_skb->len + skb_tailroom(defrag_skb),
+				     DMA_TO_DEVICE);
 	return ret;
 }
 
@@ -1774,9 +1781,9 @@ ath12k_wifi7_dp_process_rx_err_buf(struct ath12k_pdev_dp *dp_pdev,
 	list_add_tail(&desc_info->list, used_list);
 
 	rxcb = ATH12K_SKB_RXCB(msdu);
-	dma_unmap_single(ab->dev, rxcb->paddr,
-			 msdu->len + skb_tailroom(msdu),
-			 DMA_FROM_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, rxcb->paddr,
+				     msdu->len + skb_tailroom(msdu),
+				     DMA_FROM_DEVICE);
 
 	if (drop) {
 		dev_kfree_skb_any(msdu);
@@ -2322,9 +2329,9 @@ int ath12k_wifi7_dp_rx_process_wbm_err(struct ath12k_dp *dp,
 		list_add_tail(&desc_info->list, &rx_desc_used_list[device_id]);
 
 		rxcb = ATH12K_SKB_RXCB(msdu);
-		dma_unmap_single(partner_dp->dev, rxcb->paddr,
-				 msdu->len + skb_tailroom(msdu),
-				 DMA_FROM_DEVICE);
+		ath12k_core_dma_unmap_single(partner_dp->dev, rxcb->paddr,
+					     msdu->len + skb_tailroom(msdu),
+					     DMA_FROM_DEVICE);
 
 		num_buffs_reaped[device_id]++;
 		total_num_buffs_reaped++;
@@ -2470,7 +2477,7 @@ int ath12k_wifi7_dp_alloc_reo_qdesc(struct ath12k_base *ab,
 		return -ENOMEM;
 
 	*addr_aligned = PTR_ALIGN(vaddr, HAL_LINK_DESC_ALIGN);
-
+#ifdef CONFIG_IO_COHERENCY
 	paddr = dma_map_single(ab->dev, *addr_aligned, hw_desc_sz,
 			       DMA_BIDIRECTIONAL);
 	ret = dma_mapping_error(ab->dev, paddr);
@@ -2478,7 +2485,13 @@ int ath12k_wifi7_dp_alloc_reo_qdesc(struct ath12k_base *ab,
 		kfree(vaddr);
 		return ret;
 	}
-
+#else
+	paddr = virt_to_phys(*addr_aligned);
+	if (!paddr) {
+		kfree(vaddr);
+		return ret;
+	}
+#endif
 	rx_tid->vaddr = vaddr;
 	rx_tid->paddr = paddr;
 	rx_tid->size = hw_desc_sz;
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c b/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c
index eb1cf3b..75c5f98 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp_tx.c
@@ -273,6 +273,7 @@ tcl_ring_sel:
 		hdr = (void *)skb->data;
 	}
 map:
+#ifndef CONFIG_IO_COHERENCY
 	ti.paddr = dma_map_single(dp->dev, skb->data, skb->len, DMA_TO_DEVICE);
 	if (dma_mapping_error(dp->dev, ti.paddr)) {
 		atomic_inc(&dp->device_stats.tx_err.misc_fail);
@@ -280,7 +281,15 @@ map:
 		ret = -ENOMEM;
 		goto fail_remove_tx_buf;
 	}
-
+#else
+	ti.paddr = virt_to_phys(skb->data);
+	if (!ti.paddr) {
+		atomic_inc(&dp->device_stats.tx_err.misc_fail);
+		ath12k_warn(ab, "failed to DMA map data Tx buffer\n");
+		ret = -ENOMEM;
+		goto fail_remove_tx_buf;
+	}
+#endif
 	if ((!test_bit(ATH12K_FLAG_HW_CRYPTO_DISABLED, &ab->dev_flags) &&
 	     !(skb_cb->flags & ATH12K_SKB_HW_80211_ENCAP) &&
 	     !(skb_cb->flags & ATH12K_SKB_CIPHER_SET) &&
@@ -326,13 +335,17 @@ skip_htt_metadata:
 				goto fail_free_ext_skb;
 			}
 		}
-
+#ifndef CONFIG_IO_COHERENCY
 		ti.paddr = dma_map_single(dp->dev, skb_ext_desc->data,
 					  skb_ext_desc->len, DMA_TO_DEVICE);
 		ret = dma_mapping_error(dp->dev, ti.paddr);
 		if (ret)
 			goto fail_free_ext_skb;
-
+#else
+		ti.paddr = virt_to_phys(skb_ext_desc->data);
+		if (!ti.paddr)
+			goto fail_free_ext_skb;
+#endif
 		ti.data_len = skb_ext_desc->len;
 		ti.type = HAL_TCL_DESC_TYPE_EXT_DESC;
 
@@ -397,15 +410,15 @@ skip_htt_metadata:
 
 fail_unmap_dma_ext:
 	if (skb_cb->paddr_ext_desc)
-		dma_unmap_single(dp->dev, skb_cb->paddr_ext_desc,
-				 skb_ext_desc->len,
-				 DMA_TO_DEVICE);
+		ath12k_core_dma_unmap_single(dp->dev, skb_cb->paddr_ext_desc,
+					     skb_ext_desc->len,
+					     DMA_TO_DEVICE);
 fail_free_ext_skb:
 	if (skb_ext_desc)
 		kfree_skb(skb_ext_desc);
 
 fail_unmap_dma:
-	dma_unmap_single(dp->dev, ti.paddr, ti.data_len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(dp->dev, ti.paddr, ti.data_len, DMA_TO_DEVICE);
 
 fail_remove_tx_buf:
 	ath12k_dp_tx_release_txbuf(dp, tx_desc, pool_id);
@@ -433,10 +446,10 @@ static void ath12k_wifi7_dp_tx_free_txbuf(struct ath12k_dp *dp,
 	skb_cb = ATH12K_SKB_CB(msdu);
 	dp_pdev = dp->dp_pdevs[pdev_id];
 
-	dma_unmap_single(dp->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(dp->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
 	if (skb_cb->paddr_ext_desc) {
-		dma_unmap_single(dp->dev, skb_cb->paddr_ext_desc,
-				 skb_ext_desc->len, DMA_TO_DEVICE);
+		ath12k_core_dma_unmap_single(dp->dev, skb_cb->paddr_ext_desc,
+					     skb_ext_desc->len, DMA_TO_DEVICE);
 		dev_kfree_skb_any(skb_ext_desc);
 	}
 
@@ -476,10 +489,10 @@ ath12k_wifi7_dp_tx_htt_tx_complete_buf(struct ath12k_base *ab,
 	if (atomic_dec_and_test(&ar->dp.num_tx_pending))
 		wake_up(&ar->dp.tx_empty_waitq);
 
-	dma_unmap_single(ab->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
 	if (skb_cb->paddr_ext_desc) {
-		dma_unmap_single(ab->dev, skb_cb->paddr_ext_desc,
-				 skb_ext_desc->len, DMA_TO_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, skb_cb->paddr_ext_desc,
+					     skb_ext_desc->len, DMA_TO_DEVICE);
 		dev_kfree_skb_any(skb_ext_desc);
 	}
 
@@ -734,10 +747,10 @@ static void ath12k_wifi7_dp_tx_complete_msdu(struct ath12k_pdev_dp *dp_pdev,
 			dp_pdev->wmm_stats.total_wmm_tx_drop[dp_pdev->wmm_stats.tx_type]++;
 	}
 
-	dma_unmap_single(ab->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(ab->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
 	if (skb_cb->paddr_ext_desc) {
-		dma_unmap_single(ab->dev, skb_cb->paddr_ext_desc,
-				 skb_ext_desc->len, DMA_TO_DEVICE);
+		ath12k_core_dma_unmap_single(ab->dev, skb_cb->paddr_ext_desc,
+					     skb_ext_desc->len, DMA_TO_DEVICE);
 		dev_kfree_skb_any(skb_ext_desc);
 	}
 
@@ -896,12 +909,14 @@ void ath12k_wifi7_dp_tx_completion_handler(struct ath12k_dp *dp, int ring_id)
 	struct hal_wbm_release_ring *desc;
 	u8 pdev_id;
 	u64 desc_va;
+#ifndef CONFIG_IO_COHERENCY
 	int valid_entries;
-
+#endif
 	spin_lock_bh(&status_ring->lock);
 
 	ath12k_hal_srng_access_begin(ab, status_ring);
 
+#ifndef CONFIG_IO_COHERENCY
 	valid_entries = ath12k_hal_srng_dst_num_free(ab, status_ring, false);
 	if (!valid_entries) {
 		ath12k_hal_srng_access_end(ab, status_ring);
@@ -910,7 +925,7 @@ void ath12k_wifi7_dp_tx_completion_handler(struct ath12k_dp *dp, int ring_id)
 	}
 
 	ath12k_hal_srng_dst_invalidate_entry(ab, status_ring, valid_entries);
-
+#endif
 	while (ATH12K_TX_COMPL_NEXT(tx_ring->tx_status_head) != tx_ring->tx_status_tail) {
 		desc = ath12k_hal_srng_dst_get_next_cached_entry(ab, status_ring);
 		if (!desc)
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c b/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c
index 6dfb0c9..b18f417 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hal_rx.c
@@ -1164,7 +1164,7 @@ ath12k_wifi7_hal_rx_fst_attach(struct ath12k_base *ab)
 	ath12k_dbg(ab, ATH12K_DBG_DP_FST, "HAL FST allocation %pK entries %u entry size %u alloc_size %u\n",
 		   fst, fst->max_entries, fst->fst_entry_size, alloc_size);
 
-	fst->base_vaddr = dma_alloc_coherent(ab->dev, alloc_size,
+	fst->base_vaddr = ath12k_core_dma_alloc_coherent(ab->dev, alloc_size,
 					     &fst->base_paddr, GFP_KERNEL);
 	if (!fst->base_vaddr) {
 		kfree(fst);
@@ -1190,9 +1190,9 @@ void ath12k_wifi7_hal_rx_fst_detach(struct ath12k_base *ab, struct hal_rx_fst *f
 		return;
 
 	if (fst->base_vaddr)
-		dma_free_coherent(ab->dev,
-				  (fst->max_entries * fst->fst_entry_size),
-				  fst->base_vaddr, fst->base_paddr);
+		ath12k_core_dma_free_coherent(ab->dev,
+					      (fst->max_entries * fst->fst_entry_size),
+					      fst->base_vaddr, fst->base_paddr);
 	kfree(fst);
 }
 
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index 5e9c097..94f933d 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -6741,7 +6741,7 @@ static int wmi_process_mgmt_tx_comp(struct ath12k *ar, u32 desc_id,
 	spin_unlock_bh(&ar->txmgmt_idr_lock);
 
 	skb_cb = ATH12K_SKB_CB(msdu);
-	dma_unmap_single(ar->ab->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
+	ath12k_core_dma_unmap_single(ar->ab->dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
 
 	hdr = (struct ieee80211_hdr *)msdu->data;
 
-- 
2.34.1

